tsnake/snake.py

```diff
@@ -52,59 +80,147 @@ def intersects_grid_cell(self, grid_cell):
 class TSnake(object):
     """
     A class representing a *single* T-snake at a given instance in time.
-    
+
     If the initial T-snake splits into two, then there should be *two*
       instances of T-snakes present in the algorithm.
       (NOTE: can change this if inconvenient)
-      
+
     The merging two T-snakes in the algorithm should be done with:
       TSnake.merge(snake1, snake2)
-      
+
     In this class, 
         each element/edge       self.elements[i] 
             corresponds to:
         node pair               (self.nodes[i], self.nodes[i+1])
-    
+
     Args:
     ===========================================
     (list) nodes:
     * A list of Node instances, in order i=0, 1, 2, ..., N-1
     ===========================================
-    
+
     TODO: hyperparameters in constructor
     TODO: calculate intensity normal thingy
     TODO: comments for a,b, gamma
     """
-    
+
     def __init__(self, nodes, force, intensity, a, b, gamma, dt):
         for n in nodes:
             assert isinstance(n, Node)
         self.nodes = list(nodes)
         # Force and intensity fields over the image, (n,m) np arrays
         self.force = force
         self.intensity = intensity
-        #Deformation parameters
+        # Deformation parameters
         self.a = a
         self.b = b
         self.gamma = gamma
         self.dt = dt
-        
+
         self._elements = []
+        
         # Connect each node[i] --> node[i+1]
         for i in range(len(nodes)-1):
             self._elements.append(Element(self.nodes[i], self.nodes[i+1]))
-            
+
         # Connect node[N-1] --> node[0]
         self._elements.append(Element(self.nodes[-1], self.nodes[0]))
+        self._compute_normals()
+
+    def _compute_normals(self):
+        """
+        Compute normals
+        TODO: Finish doc, see if we can optimize this,
+        Computation is O(n) in the number of edges
+        Returns np array of length(# nodes, 2) representing the normal at each node
+        """
+
+        for i in range(len(self._elements)):
+            first_element = self._elements[i]
+            # Perpendicular of current element, normalized
+            norm = first_element.get_perpendicular()
+            norm /= np.sum(np.abs(norm))
```

Comment:
This is the L1 norm (divide by sum of absolute values). Any benefit of using this norm vs L2 norm (sum of squares)? Just wondering...

Reply:
We avoid a square root which is more computationally expensive. We need these to be unit vectors, i.e. length 1.